File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        stack<int> stack;
        stack.push(-1);
        int ma=0;
        for(int i=0;i<heights.size();i++){
            while(stack.top()!=-1 && heights[i]<=heights[stack.top()]){
                int height=heights[stack.top()];
                stack.pop();
                int width=i-stack.top()-1;
                ma=max(ma,height*width);
            }
            stack.push(i);
        }
        while(stack.top()!=-1){
            int height=heights[stack.top()];
            stack.pop();
            int width=heights.size()-stack.top()-1;
            ma=max(ma,height*width);
        }
        return ma;
    }
};

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.java
class Solution {
    public int largestRectangleArea(int[] arr) {
        Stack<Integer> stack = new Stack<>();
        int n = arr.length;
        int maxArea = 0;
        if(arr.length==1){
            return arr[0];
        }
        for(int index = 0; index <= n; index++) {
            int element = (index == n) ? 0 : arr[index];
            while(!stack.isEmpty() && arr[stack.peek()] > element) {
                int height = arr[stack.pop()];
                int previousSmaller = (stack.isEmpty()) ? -1 : stack.peek();
                int width = index - previousSmaller - 1;
                maxArea = Math.max(maxArea, height * width);
            }
            stack.push(index);
        }
        return maxArea;
    }
}